browser/src/statechart.js

```diff
@@ -0,0 +1,615 @@
+          target: 'overlay',
+          actions: 'updateCanvas'
+        }
+      }
+    }
+  }
+}
+
+const frameState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        'keydown.a': { actions: 'decrementFrame' },
+        'keydown.left': { actions: 'decrementFrame' },
+        'keydown.d': { actions: 'incrementFrame' },
+        'keydown.right': { actions: 'incrementFrame' },
+        'keydown.c': { actions: 'incrementChannel' },
+        'keydown.C': { actions: 'decrementChannel' },
+        'keydown.f': { actions: 'incrementFeature' },
+        'keydown.F': { actions: 'decrementFeature' },
+        LOADING: 'loading',
+      }
+    },
+    loading: {
+      on: { LOADED: 'idle' }
+    }
+  },
+  on: {
+    SETFRAME: { actions: forwardTo('backend') },
+  }
+};
+
+const confirmState = {
+  initial: 'idle',
+  states: {
+    idle: {
+      on: {
+        'keydown.o': 'predictFrame',
+        'keydown.O': 'predictAll',
+        'keydown.s': 'swapFrame',
+        'keydown.r': 'replaceFrame',
+        'keydown.R': 'replaceAll',
+      },
+    },
+    predictFrame: {
+      on: {
+        'keydown.Enter': { actions: 'predictFrame' },
+        'keydown.Escape': 'idle',
+      }
+    },
+    predictAll: {
+      on: {
+        'keydown.Enter': { actions: 'predictAll' },
+        'keydown.Escape': 'idle',
+      }
+    },
+    swapFrame: {
+      on: {
+        'keydown.Enter': { actions: 'swapFrame' },
+        'keydown.Escape': 'idle',
+      }
+    },
+    replaceFrame: {
+      on: {
+        'keydown.Enter': { actions: 'replaceFrame' },
+        'keydown.Escape': 'idle',
+      }
+    },
+    replaceAll: {
+      on: {
+        'keydown.Enter': { actions: 'replaceAll' },
+        'keydown.Escape': 'idle',
+      }
+    },
+    loading: {},
+  },
+  on: {
+    LOADING: '.loading',
+    LOADED: '.idle',
+  }
+};
+
+const rgbState = {
+  initial: 'oneChannel',
+  states: {
+    oneChannel: {
+      on: {
+        TOGGLERGB: {
+          target: 'rgb',
+          actions: send('TOGGLERGB', { to: 'backend' })
+        },
+      }
+    },
+    rgb: {
+      on: {
+        TOGGLERGB: {
+          target: 'oneChannel',
+          actions: send('TOGGLERGB', { to: 'backend' })
+        },
+      }
+    },
+  }
+};
+
+/**
+ * Handles selecting labels.
+ */
+const selectState = {
+  on: {
+    mousedown: { actions: 'selectLabel' },
+    'keydown.x': { actions: 'swapLabels' },
+    'keydown.n': { actions: 'resetLabels' },
+    'keydown.[': { actions: 'decrementForeground' },
+    'keydown.]': { actions: 'incrementForeground' },
+    'keydown.{': { actions: 'decrementBackground' },
+    'keydown.}': { actions: 'incrementBackground' },
+  }
+};
+
+export const deepcellLabelMachine = Machine(
+  {
+    id: 'deepcellLabel',
+    type: 'parallel',
+    context: {
+      trace: [],
+      storedLabel: 0,
+      storedX: 0,
+      storedY: 0,
+    },
+    invoke: [
+      backendMachine,
+    ],
+    states: {
+      rgb: rgbState,
```

Comment:
<img width="769" alt="Screen Shot 2021-02-16 at 2 18 03 PM" src="https://user-images.githubusercontent.com/40068392/108130572-bb849180-7064-11eb-83a6-860d0227b30b.png">

These first states are more straightforward, with most of them simply carrying out actions in response to events.

The rgb state also tracks whether we are viewing an RGB image, and the canvas state tracks whether the space bar is down & we are panning. As these states are [orthogonal or parallel
](https://xstate.js.org/docs/guides/parallel.html), we are in all of these states simultaneously and all the actions are always available (i.e. we can always change the brush size because we are always in the brush state).